File: linked_lists/fast_slow_pointers/python3/delete_middle_node.py
from node import ListNode
from singly import SinglyLinkedList

def delete_middle_node(head: ListNode) -> ListNode:
    """
    Given the `head` of a singly linked list, delete the middle node.

    Parameters
    ----------
    head : ListNode
        The sentinel head of the linked list

    Returns
    -------
    ListNode
        The sentinel head of the modified linked list with the middle node removed
    """
    # Handle single node (i.e., if the first real node is the only node)
    if head.next.next is None:
        # Set the first real node after the sentinel head to None to remove it
        head.next = None
        return head
    
    # Initialize pointers to the first real node
    slow, fast, prev = head.next, head.next, head.next

    while fast and fast.next:
        fast = fast.next.next
        # Previous points to the node right before slow
        prev = slow
        slow = slow.next

    # Slow points to the middl node and prev points to the node before it, remove slow
    prev.next = slow.next

    return head

def main() -> int:

    test_cases = [[1, 3, 4, 7, 1, 2, 6], [1, 2, 3, 4], [2, 1], [1]]

    for data_array in test_cases:
        nodes = (ListNode(data=data) for data in data_array)
        sll = SinglyLinkedList()
        for node in nodes:
            sll.add_to_end(node)
        print("Before removing the middle node:")
        sll.display()
        head = delete_middle_node(sll.head)
        print("After removing the middle node:")
        sll.display()
        print('\n')

    return 0

if __name__ == "__main__":

    main()

File: linked_lists/fast_slow_pointers/c++/delete_middle_node/delete_middle_node.cpp
// delete_middle_node.cpp
#include "delete_middle_node.hpp"
#include "node.hpp"

ListNode::Ptr deleteMiddleNode(const ListNode::Ptr &head)
{
    // Single node after sentinel head
    if (head->next->next == nullptr)
    {
        // Remove the single node as it is considered the middle node
        head->next = nullptr;
        return head;
    }

    // Shared pointers instead of raw pointers
    ListNode::Ptr fast = head->next;
    ListNode::Ptr slow = head->next;
    ListNode::Ptr prev = head->next;

    while (fast != nullptr && fast->next != nullptr)
    {
        fast = fast->next->next;
        // Set prev to slow before iterating slow to the next node
        prev = slow;
        slow = slow->next;
    }

    // Remove the node pointed to by the slow pointer
    prev->next = slow->next;

    return head;
}
